### DISCUSSION TITLE
Announcing the Appwrite .NET SDK (Beta) 🚀

### DISCUSSION THREAD

**Discussion by adityaoberai** [appwrite/appwrite#5820] [Ideas]
We're excited to share that we now have a [.NET SDK](https://github.com/appwrite/sdk-for-dotnet) available to build with Appwrite. This is a server-side SDK that targets .NET Standard 2.0, supporting a wide range of SDK versions across .NET, .NET Core, and .NET Framework.

This SDK is currently in beta, and we hope to improve it over the coming weeks. We would highly appreciate it if you would try it out and share your suggestions and feedback.

## Installation

You can access the SDK via the [NuGet Package Manager](https://www.nuget.org/packages/Appwrite/0.4.2).

To use it in your .NET project, add this reference to your project's `.csproj` file:

```xml
<PackageReference Include="Appwrite" Version="0.4.2" />
```

You can also install the package from the command line:

```bash
# Package Manager
Install-Package Appwrite -Version 0.4.2

# or .NET CLI
dotnet add package Appwrite --version 0.4.2
```

### Initialize The SDK

Once you have installed the package, it is extremely easy to get started with the SDK; simply import the package in your code and set your Appwrite credentials to start making API calls.

```csharp
using Appwrite;

var client = new Client()
  .SetEndpoint("http://cloud.appwrite.io/v1")
  .SetProject("5ff3379a01d25")
  .SetKey("cd868db89");
```

### Make API Requests

Once your SDK is initialized, you can start making API calls. Here are some examples:

### Create A User

```csharp
using Appwrite;
using Appwrite.Services;
using Appwrite.Models;

var users = new Users(client);

var user = await users.Create(
    userId: ID.Unique(),
    email: "email@example.com",
    password: "password",
    name: "name");
```

### Add A Document

```csharp
using Appwrite;
using Appwrite.Services;
using Appwrite.Models;

var databases = new Databases(client);

var document = await databases.CreateDocument(
    databaseId: "[DATABASE_ID]",
    collectionId: "[COLLECTION_ID]",
    documentId: "[DOCUMENT_ID]",
    data: [object]);
```

### Upload A File

```csharp
using Appwrite;
using Appwrite.Services;
using Appwrite.Models;

var storage = new Storage(client);

var file = await storage.CreateFile(
    bucketId: "[BUCKET_ID]",
    fileId: "[FILE_ID]",
    file: new File("./path-to-files/image.jpg"));
```

### Trigger A Function Execution

```csharp
using Appwrite;
using Appwrite.Services;
using Appwrite.Models;

var functions = new Functions(client);

var execution = await functions.CreateExecution(
    functionId: "[FUNCTION_ID]");
```

### Error Handling

The Appwrite .NET SDK throws an `AppwriteException` object which includes `message`, `code`, and `response` properties. You can handle any errors by catching `AppwriteException` and presenting the `message` to the user or using the [error types](https://appwrite.io/docs/response-codes#errorTypes) provided by Appwrite to decide the right type of error and display custom error messages. Below is an example.

```csharp
using Appwrite;
using Appwrite.Services;
using Appwrite.Models;

var users = new Users(client);

try
{
    var user = await users.Create(
        userId: ID.Unique(),
        email: "email@example.com",
        password: "password",
        name: "name");
}
catch (AppwriteException e)
{
    Console.WriteLine(e.Message);
}
```

You can also visit our [Playground for .NET repo](https://github.com/appwrite/playground-for-dotnet) and clone it to try out the .NET SDK in a test project.

## Contributing and Sharing Feedback

In case you discover any bugs/issues, want to share suggestions, or make contributions to the .NET SDK, please visit Appwrite’s [SDK Generator](https://github.com/appwrite/sdk-generator) repository on GitHub.

All contributions, including those by people with commit access, must go through a pull request and be approved by an Appwrite core team member before merging. This is to ensure a proper review of all the contributions. You can learn more by visiting the repo’s [Contribution Guide](https://github.com/appwrite/sdk-generator/blob/master/CONTRIBUTING.md).

## Further Steps

We hope you will try the .NET SDK in your upcoming projects and share any thoughts and concerns you have with us. Here are some resources you can use to learn more about Appwrite and start building projects:

- 📜 [Appwrite Docs](https://appwrite.io/docs)
- 💬 [Discord Community](https://appwrite.io/discord)
- 📺 [Appwrite YouTube Channel](https://www.youtube.com/@Appwrite)

**Comment by christyjacob4** [comment]
@adityaoberai This is awesome news! 👏  I think I should start learning a bit of .NET now. I'm still not at terms with the style guide for brackets 😆  but I will just rely on the linter for that

**Reply by adityaoberai** [reply]
Haha, yes 🤣

**Reply by AathifMahir** [reply]
I guess as you move forward with dotnet, you'd understand how broad and flexible the dotnet ecosystem is.

**Reply by adityaoberai** [reply]
Oh it's an absolutely wonderful world! 😁

One at least I thoroughly enjoy being a part of ❤️

**Comment by ashishdawale20** [comment]
Hello @adityaoberai ,
One small observation, the users.List() method is just giving back top 25 users while it correctly returns the total users count.
Is that the default behavior or bug? 

If that is the default Behavior, how to fetch next 25 users?

**Reply by adityaoberai** [reply]
25 is the default limit set to prevent the Appwrite instance from abuse, but you can change the limit using `Query.Limit()` and then implement offset pagination using `Query.Offset()` 

https://appwrite.io/docs/pagination

**Reply by ashishdawale20** [reply]
Thank you for the quick reply

**Comment by AathifMahir** [comment]
Any news on Client SDK, Dotnet has lots of app models and framework that could take advantage of client SDK and game changer for many devs who need platform like appwrite.

Currently dotnet has lots of different frameworks for building native, cross platform and web applications, eg: WinUI, UWP, WPF, Maui, Avalonia, Uno, Blazor Wasm, Unity, etc.. @adityaoberai

**Reply by adityaoberai** [reply]
Now that the server-side .NET SDK is in a healthy state, we'll definitely be looking towards client-side SDKs.

Based on existing community demand, the first one we'll be focusing on is for Unity, and we'll explore other frameworks post that.

**Reply by AathifMahir** [reply]
> Now that the server-side .NET SDK is in a healthy state, we'll definitely be looking towards client-side SDKs.
> 
> Based on existing community demand, the first one we'll be focusing on is for Unity, and we'll explore other frameworks post that.

That's awesome to hear, is the client SDK is based on old client SDK which has been under maintenance state for while or completely new one that was written from scratch

**Reply by adityaoberai** [reply]
> > Now that the server-side .NET SDK is in a healthy state, we'll definitely be looking towards client-side SDKs.
> > 
> > Based on existing community demand, the first one we'll be focusing on is for Unity, and we'll explore other frameworks post that.
> 
> That's awesome to hear, is the client SDK is based on old client SDK which has been under maintenance state for while or completely new one that was written from scratch

We haven't had any Client SDKs in the past, but we'll be using the current server SDK as inspiration since a major chunk of the services would work in a similar manner.

Definitely have some exploring to do here, but that we shall make sure to invest some time towards :)

**Reply by AathifMahir** [reply]
> > > Now that the server-side .NET SDK is in a healthy state, we'll definitely be looking towards client-side SDKs.
> > > 
> > > Based on existing community demand, the first one we'll be focusing on is for Unity, and we'll explore other frameworks post that.
> > 
> > That's awesome to hear, is the client SDK is based on old client SDK which has been under maintenance state for while or completely new one that was written from scratch
> 
> We haven't had any Client SDKs in the past, but we'll be using the current server SDK as inspiration since a major chunk of the services would work in a similar manner :)

If I'm not mistaken there's an archived client appwrite SDK for dotnet in the past. Any that's great to hear since both of them are mostly the same

**Reply by adityaoberai** [reply]
> > > > Now that the server-side .NET SDK is in a healthy state, we'll definitely be looking towards client-side SDKs.
> > > > 
> > > > Based on existing community demand, the first one we'll be focusing on is for Unity, and we'll explore other frameworks post that.
> > > 
> > > That's awesome to hear, is the client SDK is based on old client SDK which has been under maintenance state for while or completely new one that was written from scratch
> > 
> > We haven't had any Client SDKs in the past, but we'll be using the current server SDK as inspiration since a major chunk of the services would work in a similar manner :)
> 
> If I'm not mistaken there's an archived client appwrite SDK for dotnet in the past. Any that's great to hear since both of them are mostly the same

There may have been certain community-built libraries for .NET client frameworks but officially, these will be our first efforts in this direction 😁
